File: hot100/Solution0437.cpp
#include <unordered_map>
using namespace std;
// 给定一个二叉树的根节点 root ，和一个整数 targetSum ，求该二叉树里节点值之和等于 targetSum 的 路径 的数目。

// 路径 不需要从根节点开始，也不需要在叶子节点结束，但是路径方向必须是向下的（只能从父节点到子节点）。

// 示例 1：

// 输入：root = [10,5,-3,3,2,null,11,3,-2,null,1], targetSum = 8
// 输出：3
// 解释：和等于 8 的路径有 3 条，如图所示。
// 示例 2：

// 输入：root = [5,4,8,11,null,13,4,7,2,null,null,5,1], targetSum = 22
// 输出：3

// 提示:

// 二叉树的节点个数的范围是 [0,1000]
// -109 <= Node.val <= 109
// -1000 <= targetSum <= 1000
// 方法一：前序遍历每个节点，每个节点再作为路径头递归
class Solution
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

    int count = 0;
    long long target = 0;

public:
    int pathSum(TreeNode *root, int targetSum)
    {
        target = targetSum;
        preOrder(root);
        return count;
    }

private:
    void preOrder(TreeNode *node)
    {
        if (!node)
            return;
        addToPath(node, 0);
        preOrder(node->left);
        preOrder(node->right);
    }

    void addToPath(TreeNode *node, long long currentSum)
    {
        if (!node)
            return;
        currentSum += node->val;
        if (currentSum == target)
            count++;
        // 从当前节点开始的路径
        addToPath(node->left, currentSum);
        addToPath(node->right, currentSum);
    }
};
// 方法二：每个节点作为路径头+左节点作为中途点+右节点作为中途点
class Solution2
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

public:
    int pathSum(TreeNode *root, int targetSum)
    {
        if (!root)
            return 0;
        int result = rootSum(root, (long long) targetSum);
        result += pathSum(root->left, targetSum);
        result += pathSum(root->right, targetSum);
        return result;
    }

private:
    int rootSum(TreeNode *node, long long targetSum)
    {
        if (!node)
            return 0;
        int count = 0;
        if (node->val == targetSum)
            count++;
        count += rootSum(node->left, targetSum - node->val);
        count += rootSum(node->right, targetSum - node->val);
        return count;
    }
};

// 方法三：前缀和+哈希表。
// 假设现在遍历到了current，而存在某一个节点q，使得q到current的路径和为targetSum。
// (根到current的前缀和) - targetSum = (根到q的前缀和)
// 我们并不关心q具体在哪？我们只关心存在多少个(根到q的前缀和)
class Solution3
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

    unordered_map<long long, int> prefixSumCount; // 存储前缀和及其出现次数
    int target;

public:
    int pathSum(TreeNode *node, int targetSum)
    {
        target = targetSum;
        prefixSumCount[0] = 1; // 前缀和为0出现一次
        return dfs(node, 0);
    }

private:
    int dfs(TreeNode *node, long long currentSum)
    {
        if (!node)
            return 0;

        currentSum += node->val; // 更新当前前缀和

        int result = prefixSumCount.count(currentSum - target) ? prefixSumCount[currentSum - target] : 0; // 查找当前前缀和减去目标值的次数，即有多少个p

        prefixSumCount[currentSum]++; // 记录当前前缀和出现次数，这一步不能出现在result之前

        result += dfs(node->left, currentSum); // 递归左子树
        result += dfs(node->right, currentSum); // 递归右子树

        prefixSumCount[currentSum]--; // 回溯，撤销当前前缀和的记录。因为一个节点访问完了其所有子节点，则要回其父节点，父节点再去访问右子树，因此，这个节点不会再出现在他父亲的路线中了
        return result;
    }
};

File: hot100/Solution0437.java
package hot100;

import java.util.HashMap;
import java.util.Map;
// 给定一个二叉树的根节点 root ，和一个整数 targetSum ，求该二叉树里节点值之和等于 targetSum 的 路径 的数目。

// 路径 不需要从根节点开始，也不需要在叶子节点结束，但是路径方向必须是向下的（只能从父节点到子节点）。

public class Solution0437 {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return 0;
        }
        return pathSumFrom(root, (long) targetSum) + pathSum(root.left, targetSum) + pathSum(root.right, targetSum);
    }

    public int pathSumFrom(TreeNode node, long targetSum) {
        if (node == null) {
            return 0;
        }
        int count = 0;
        if (node.val == targetSum) {
            count++;
        }
        count += pathSumFrom(node.left, targetSum - node.val); // 从左子节点开始的路径数
        count += pathSumFrom(node.right, targetSum - node.val); // 从右子节点开始的路径数
        return count;
    }

    private int count2 = 0;
    private long target;

    public int pathSum2(TreeNode root, int targetSum) {
        if (root == null) {
            return 0;
        }
        target = targetSum;
        dfs2(root);
        return count2;
    }

    private void dfs2(TreeNode node) {
        if (node == null) {
            return;
        }
        addToPath(node, 0);
        dfs2(node.left);
        dfs2(node.right);
    }

    private void addToPath(TreeNode node, long currentSum) {
        if (node == null) {
            return;
        }
        currentSum += node.val;
        if (currentSum == target) {
            count2++;
        }
        addToPath(node.left, currentSum);
        addToPath(node.right, currentSum);
    }

    private int count3 = 0;

    public int pathSum3(TreeNode root, int targetSum) {
        if (root == null) {
            return 0;
        }
        dfs3(root, (long) targetSum);
        pathSum3(root.left, targetSum);
        pathSum3(root.right, targetSum);
        return count3;
    }

    private void dfs3(TreeNode node, long targetSum) { // 以当前节点为路径头的路径数
        if (node == null) {
            return;
        }
        if (node.val == targetSum) {
            count3++;
        }
        dfs3(node.left, targetSum - node.val);
        dfs3(node.right, targetSum - node.val);
    }

    private Map<Long, Integer> prefixSumCount = new HashMap<>();
    private int count4 = 0;

    public int pathSum4(TreeNode root, int targetSum) {
        prefixSumCount.put(0L, 1);
        dfs4(root, 0L, (long) targetSum);
        return count4;
    }

    private void dfs4(TreeNode node, long currentSum, long targetSum) {
        if (node == null) {
            return;
        }
        currentSum += node.val;
        count4 += prefixSumCount.getOrDefault(currentSum - targetSum, 0);
        prefixSumCount.put(currentSum, prefixSumCount.getOrDefault(currentSum, 0) + 1); // 将当前前缀和加入哈希表
        dfs4(node.left, currentSum, targetSum);
        dfs4(node.right, currentSum, targetSum);
        prefixSumCount.put(currentSum, prefixSumCount.get(currentSum) - 1); // 递归结束后是下一个节点的前缀和，当前前缀和不再需要，减少计数
    }

    public static void main(String[] args) {
        Solution0437 solution = new Solution0437();
        TreeNode root = new TreeNode(10);
        root.left = new TreeNode(5);
        root.right = new TreeNode(-3);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(2);
        root.right.right = new TreeNode(11);
        root.left.left.left = new TreeNode(3);
        root.left.left.right = new TreeNode(-2);
        root.left.right.right = new TreeNode(1);
        int targetSum = 8;
        int result = solution.pathSum(root, targetSum);
        System.out.println(result); // 输出: 3

        int result2 = solution.pathSum2(root, targetSum);
        System.out.println(result2); // 输出: 3

        int result3 = solution.pathSum3(root, targetSum);
        System.out.println(result3); // 输出: 3

        int result4 = solution.pathSum4(root, targetSum);
        System.out.println(result4); // 输出: 3
    }
}
